core/providers/dotnet/dotnet.go

```diff
@@ -0,0 +1,205 @@
+package dotnet
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"maps"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/railwayapp/railpack/core/generate"
+	"github.com/railwayapp/railpack/core/plan"
+	"github.com/railwayapp/railpack/internal/utils"
+)
+
+const (
+	DEFAULT_DOTNET_VERSION   = "6.0.428"
+	DOTNET_ROOT              = "/mise/installs/dotnet"
+	DOTNET_DEPENDENCIES_ROOT = "/root/.nuget/packages"
+)
+
+type DotnetProvider struct {
+}
+
+func (p *DotnetProvider) Name() string {
+	return "Dotnet"
+}
+
+func (p *DotnetProvider) Detect(ctx *generate.GenerateContext) (bool, error) {
+	return ctx.App.HasMatch("*.csproj"), nil
+}
+
+func (p *DotnetProvider) Initialize(ctx *generate.GenerateContext) error {
+	return nil
+}
+
+func (p *DotnetProvider) Plan(ctx *generate.GenerateContext) error {
+	miseStep := ctx.GetMiseStepBuilder()
+	p.InstallMisePackages(ctx, miseStep)
+
+	install := ctx.NewCommandStep("install")
+	install.AddInput(plan.NewStepLayer(miseStep.Name()))
+	p.Install(ctx, install)
+
+	build := ctx.NewCommandStep("build")
+	build.AddInput(plan.NewStepLayer(miseStep.Name()))
+	build.AddInput(plan.NewStepLayer(install.Name(), plan.Filter{
+		Include: []string{"obj/", DOTNET_DEPENDENCIES_ROOT},
+	}))
+	p.Build(ctx, build)
+
+	envVars := p.GetEnvVars(ctx)
+	// Required for internationalization
+	ctx.Deploy.AddAptPackages([]string{"libicu-dev"})
+	ctx.Deploy.AddInputs([]plan.Layer{
+		plan.NewStepLayer(miseStep.Name(), plan.Filter{
+			// Need to include the dotnet runtime for the binary to run
+			Include: []string{envVars["DOTNET_ROOT"]},
+		}),
+		plan.NewStepLayer(build.Name(), plan.Filter{
+			Include: []string{"out"},
+		}),
+	})
+	ctx.Deploy.StartCmd = p.GetStartCommand(ctx)
+	maps.Copy(ctx.Deploy.Variables, envVars)
+
+	return nil
+}
+
+func (p *DotnetProvider) StartCommandHelp() string {
+	return "To start your Dotnet application, Railpack will look for:\n\n" +
+		"1. A .csproj file in your project root\n\n" +
+		"The project will be run with `./out`"
+}
+
+func (p *DotnetProvider) GetStartCommand(ctx *generate.GenerateContext) string {
+	projFiles, err := ctx.App.FindFiles("*.csproj")
+	if err != nil || len(projFiles) == 0 {
+		return ""
+	}
+	projFile := projFiles[0]
+	projName := strings.TrimSuffix(projFile, ".csproj")
+	return fmt.Sprintf("./out/%s", projName)
+}
+
+func (p *DotnetProvider) Install(ctx *generate.GenerateContext, install *generate.CommandStepBuilder) {
+	maps.Copy(install.Variables, p.GetEnvVars(ctx))
+	install.AddCommands([]plan.Command{
+		plan.NewCopyCommand("nuget.config*"),
+		plan.NewCopyCommand("*.csproj"),
+		plan.NewCopyCommand("global.json*"),
+		plan.NewExecCommand(fmt.Sprintf("mkdir -p %s", DOTNET_DEPENDENCIES_ROOT)),
+		plan.NewExecCommand(`dotnet restore`),
+	})
+}
+
+func (p *DotnetProvider) Build(ctx *generate.GenerateContext, build *generate.CommandStepBuilder) {
+	maps.Copy(build.Variables, p.GetEnvVars(ctx))
+	build.AddCommands([]plan.Command{
+		plan.NewCopyCommand("."),
```

Comment:
[we should use NewLocalLayer](https://github.com/railwayapp/railpack/blob/afb0ac970de386a0b8de0a9a077d132d0ec79b7a/core/providers/python/python.go#L73) which ensures that `.dockerignore` is respected